packages/parse5-htmlparser2-tree-adapter/lib/index.ts

```diff
@@ -0,0 +1,264 @@
+        prev.next = newNode;
+        newNode.prev = prev;
+    }
+
+    referenceNode.prev = newNode;
+    newNode.next = referenceNode;
+
+    parentNode.children.splice(insertionIdx, 0, newNode);
+    newNode.parent = parentNode;
+}
+
+export function setTemplateContent(templateElement: Element, contentElement: Document): void {
+    appendChild(templateElement, contentElement);
+}
+
+export function getTemplateContent(templateElement: Element): Node {
+    return templateElement.children[0];
+}
+
+export function setDocumentType(
+    document: Document,
+    name: string | null,
+    publicId: string | null,
+    systemId: string | null
+): void {
+    const data = doctype.serializeContent(name, publicId, systemId);
+    let doctypeNode = document.children.find(
+        (node) => isDirective(node) && node.name === '!doctype'
+    ) as ProcessingInstruction;
+
+    if (doctypeNode) {
+        doctypeNode.data = data ?? null;
+    } else {
+        doctypeNode = new ProcessingInstruction('!doctype', data);
+        appendChild(document, doctypeNode);
+    }
+
+    doctypeNode['x-name'] = name ?? undefined;
+    doctypeNode['x-publicId'] = publicId ?? undefined;
+    doctypeNode['x-systemId'] = systemId ?? undefined;
+}
+
+export function setDocumentMode(document: Document, mode: DOCUMENT_MODE): void {
+    document['x-mode'] = mode;
+}
+
+export function getDocumentMode(document: Document): DOCUMENT_MODE {
+    return document['x-mode'] as DOCUMENT_MODE;
+}
+
+export function detachNode(node: Node): void {
+    if (node.parent) {
+        const idx = node.parent.children.indexOf(node);
+        const { prev, next } = node;
+
+        node.prev = null;
+        node.next = null;
+
+        if (prev) {
+            prev.next = next;
+        }
+
+        if (next) {
+            next.prev = prev;
+        }
+
+        node.parent.children.splice(idx, 1);
+        node.parent = null;
+    }
+}
+
+export function insertText(parentNode: NodeWithChildren, text: string): void {
+    const lastChild = parentNode.children[parentNode.children.length - 1];
+
+    if (lastChild && isText(lastChild)) {
+        lastChild.data += text;
+    } else {
+        appendChild(parentNode, createTextNode(text));
+    }
+}
+
+export function insertTextBefore(parentNode: NodeWithChildren, text: string, referenceNode: Node): void {
+    const prevNode = parentNode.children[parentNode.children.indexOf(referenceNode) - 1];
+
+    if (prevNode && isText(prevNode)) {
+        prevNode.data += text;
+    } else {
+        insertBefore(parentNode, createTextNode(text), referenceNode);
+    }
+}
+
+export function adoptAttributes(recipient: Element, attrs: Attribute[]): void {
+    for (let i = 0; i < attrs.length; i++) {
+        const attrName = attrs[i].name;
+
+        if (typeof recipient.attribs[attrName] === 'undefined') {
+            recipient.attribs[attrName] = attrs[i].value;
+            recipient['x-attribsNamespace']![attrName] = attrs[i].namespace!;
```

Comment:
typescript-eslint should warn on non-null assertions, it would be good if one day we can get rid of these. IIRC there's a lot of them, since the original source made a lot of assumptions about what has executed beforehand. just a note anyway, may open an issue so we don't forget